query_class_config.hh

```diff
@@ -23,19 +23,91 @@
 
 #include <cinttypes>
 
+namespace ser {
+
+template <typename T>
+class serializer;
+
+};
+
+
 namespace query {
 
+/*
+ * This struct is used in two incompatible ways.
+ *
+ * SEPARATE_PAGE_SIZE_AND_SAFETY_LIMIT cluster feature determines which way is
+ * used.
+ *
+ * 1. If SEPARATE_PAGE_SIZE_AND_SAFETY_LIMIT is not enabled on the cluster then
+ *    `page_size` field is ignored. Depending on the query type the meaning of
+ *    the remaining two fields is:
+ *
+ *    a. For unpaged queries or for reverse queries:
+ *
+ *          * `soft_limit` is used to warn about queries that result exceeds
+ *            this limit. If the limit is exceeded, a warning will be written to
+ *            the log.
+ *
+ *          * `hard_limit` is used to terminate a query which result exceeds
+ *            this limit. If the limit is exceeded, the operation will end with
+ *            an exception.
+ *
+ *    b. For all other queries, `soft_limit` == `hard_limit` and their value is
+ *       really a page_size in bytes. If the page is not previously cut by the
+ *       page row limit then reaching the size of `soft_limit`/`hard_limit`
+ *       bytes will cause a page to be finished.
+ *
+ * 2. If SEPARATE_PAGE_SIZE_AND_SAFETY_LIMIT is enabled on the cluster then all
+ *    three fields are always set. They are used in different places:
+ *
+ *    a. `soft_limit` and `hard_limit` are used for unpaged queries and in a
+ *       reversing reader used for reading KA/LA sstables. Their meaning is the
+ *       same as in (1.a) above.
+ *
+ *    b. all other queries use `page_size` field only and the meaning of the
+ *       field is the same ase in (1.b) above.
+ *
+ * Two interpretations of the `max_result_size` struct are not compatible so we
+ * need to take care of handling a mixed clusters.
+ *
+ * As long as SEPARATE_PAGE_SIZE_AND_SAFETY_LIMIT cluster feature is not
+ * supported by all nodes in the clustser, new nodes will always use the
+ * interpretation described in the point (1). `soft_limit` and `hard_limit`
+ * fields will be set appropriately to the query type and `page_size` field
+ * will be set to 0. Old nodes will ignare `page_size` anyways and new nodes
+ * will know to ignore it as well when it's set to 0. Old nodes will never set
+ * `page_size` and that means new nodes will give it a default value of 0 and
+ * ignore it for messages that miss this field.
+ *
+ * Once SEPARATE_PAGE_SIZE_AND_SAFETY_LIMIT cluster feature becomes supported by
+ * the whole cluster, new nodes will start to set `page_size` to the right value
+ * according to the interpretation described in the point (2).
```

Comment:
There's a race here.

1. a query is sent when SEPARATE_PAGE_SIZE_AND_SAFETY_LIMIT is unset
2. the cluster negotiates SEPARATE_PAGE_SIZE_AND_SAFETY_LIMIT
3. the query sent in step 1 is evaluated

The query was created with one meaning and evaluated with another. This will result in a page size of 0, and reconciliation problems.

I think the fix is as follows:

1. only the coordinator looks at SEPARATE_PAGE_SIZE_AND_SAFETY_LIMIT (and follows the rules above)
2. the replica looks at page_size. If it was unset (0), it knows the coordinator did not have SEPARATE_PAGE_SIZE_AND_SAFETY_LIMIT when it sent the query. It then uses the rules for "before". If page_size was nonzero, it knows the coordinator is using the new rules.

The coordinator must also capture the value of SEPARATE_PAGE_SIZE_AND_SAFETY_LIMIT once per query and use the captured value, so it doesn't switch rules during a query execution.

Reply:
Ah, I see it already works as I described. So the comment needs to be elaborated a little, please follow up with an update.

Reply:
Yes. I think I've got it covered. In general, it's only safe for coordinators to look at the cluster feature. Each message has to be self explanatory so that the replica could decide based only on the message. In our case, page_size == 0 or missing means we're going the old way for this request.

Reply:
I updated the comment here: https://github.com/scylladb/scylla/pull/9855

Reply:
@avikivity ^